jsapp/js/projects/projectsTable/colResize.tsx

```diff
@@ -0,0 +1,271 @@
+ *     initialization, from text widths of localized headings.
+ * - Consequence of fluid widths
+ *   - Try resizing different columns in different orders, sometimes neighbors
+ *     adapt in unexpected ways.
+ *   - Could use hard px values for defaults to avoid.
+ * - Persistence
+ *   - Widths are cleared on refresh (or unmount)
+ *   - Future, could save them in localstorage, clearing on logout.
+ * - Touch events
+ *   - Haven't tested kpi on a touchscreen recently, but the code is adaptable.
+ *   - Might be pretty simple to support by switching to pointer events.
+ */
+
+/**
+ * A flexible subset of fieldnames mapped to widths in px.
+ * Example:
+ * {
+ *   name:   100,  //  the "name" column will be 100px wide
+ *   status: 150  // the "status" column will be 150px wide
+ * }
+ */
+interface ColumnWidths {
+  [key: string]: number;
+}
+interface ColumnWidthReducerAction {
+  type: 'resize';
+  column: string;
+  width: number;
+}
+/**
+ * Reducer for column width updates.
+ * Almost too simple to justify its existence.
+ *
+ * Pulls a tiny bit of update logic out of the event handlers.
+ */
+function columnWidthReducer(
+  columnWidths: ColumnWidths,
+  action: ColumnWidthReducerAction
+) {
+  switch (action.type) {
+    case 'resize':
+      return {
+        ...columnWidths,
+        [action.column]: action.width,
+      };
+    default:
+      return {...columnWidths};
+  }
+}
+
+/**
+ * The minimum resizing width for a given column.
+ *
+ * This prevents columns from being entirely cropped away.
+ * Currently a minimum of 108px.
+ * Can be hardcoded per-field.
+ */
+function minColWidth(fieldname: string) {
+  const defaultMinimumWidth = 108;
+  const minimums: ColumnWidths = {
+    countries: 116,
+    dateModified: 144,
+    dateDeployed: 144,
+  };
+  return minimums[fieldname] || defaultMinimumWidth;
+}
+/**
+ * The maximum resizing width for a given column.
+ *
+ * This is intended to reduce confusion from resizing 1 column absurdly wide.
+ * Currently an arbitrary maximum 800px for all fields.
+ * Can be hardcoded per-field.
+ */
+function maxColWidth(fieldname: string) {
+  const defaultMaximumWidth = 800;
+  const maximums: ColumnWidths = {};
+  return maximums[fieldname] || defaultMaximumWidth;
+}
+/**
+ * A helper to clamp between the minimum and maximum allowed resizing widths.
+ */
+function clampedColumnWidth(fieldname: string, width: number) {
+  return Math.min(
+    Math.max(width, minColWidth(fieldname)),
+    maxColWidth(fieldname)
+  );
+}
+
+/**
+ * <ColumnResizer/>
+ *
+ * Creates and updates styles for column widths based on drag events on resize
+ * handles.
+ */
+export default function ColumnResizer() {
+  // State used in rendering new styles
+  const [columnWidths, dispatch] = useReducer(
+    columnWidthReducer,
+    {} as ColumnWidths
+  );
+  // Polish. This could perhaps be merged into a single state with isDragging
+  const [shouldRenderBodyCursor, setShouldRenderBodyCursor] = useState(false);
```

Comment:
Actually, we can't combine them, both are needed for reasons.

- **isDraggingRef**: Event handlers can't read up-to-date values of React state. So functionally, the event handlers need isDraggingRef for their logic, and it can't be state.
- **shouldRenderBodyCursor**, aka **isDragging** (state): We do want to trigger some re-rendering when starting/stopping a drag event, so we need state variable here.